service/forward_service.cc

```diff
@@ -42,12 +47,165 @@
+    static const db::functions::function_name COUNT_NAME = db::functions::function_name::native_function("count");
+    schema_ptr schema = local_schema_registry().get(request.cmd.schema_version);
+    std::vector<::shared_ptr<db::functions::aggregate_function>> aggrs;
+
+    auto name_as_type = [&] (const sstring& name) -> data_type {
+        return schema->get_column_definition(to_bytes(name))->type->underlying_type();
+    };
+
+    for (size_t i = 0; i < request.reductions.size(); i++) {
+        ::shared_ptr<db::functions::aggregate_function> aggr;
+
+        if (!request.aggregation_infos) {
+            if (request.reductions[i] == query::forward_request::reduction_type::aggregate) {
+                throw std::runtime_error("No aggregation info for reduction type aggregation.");
+            }
+
+            auto name = db::functions::function_name::native_function("countRows");
+            auto func = cql3::functions::functions::find(name, {});
+            aggr = dynamic_pointer_cast<db::functions::aggregate_function>(func);
+            if (!aggr) {
+                throw std::runtime_error("Count function not found.");
+            }
+        } else {
+            auto& info = request.aggregation_infos.value()[i];
+            auto types = boost::copy_range<std::vector<data_type>>(info.column_names | boost::adaptors::transformed(name_as_type));
+            auto func = cql3::functions::functions::find(info.name, types);
+
+            if (!func) {
+                if (info.name == MIN_NAME) {
+                    if (types.size() != 1) {
+                        throw std::runtime_error("min() function requires only 1 argument");
+                    }
+
+                    auto& arg = types[0];
+                    if (arg->is_collection() || arg->is_tuple() || arg->is_user_type()) {
+                        func = cql3::functions::aggregate_fcts::make_min_dynamic_function(arg);
+                    }
+
+                } else if (info.name == MAX_NAME) {
+                    if (types.size() != 1) {
+                        throw std::runtime_error("max() function requires only 1 argument");
+                    }
+
+                    auto& arg = types[0];
+                    if (arg->is_collection() || arg->is_tuple() || arg->is_user_type()) {
+                        func = cql3::functions::aggregate_fcts::make_max_dynamic_function(arg);
+                    }
+                } else if (info.name == COUNT_NAME) {
+                    if (types.size() != 1) {
+                        throw std::runtime_error("count() function requires only 1 argument");
+                    }
+
+                    auto& arg = types[0];
+                    if (arg->is_collection() || arg->is_tuple() || arg->is_user_type()) {
+                        func = cql3::functions::aggregate_fcts::make_count_rows_function();
+                    }
```

Comment:
Can't we make the functions code do all this? It seems wrong to have the service know about all the functions.

Reply:
Moved to `cql3::functions::functions::find_with_dynamic_aggr()`